File: typescript/13.罗马数字转整数.ts
/*
 * @lc app=leetcode.cn id=13 lang=typescript
 *
 * [13] 罗马数字转整数
 */

// @lc code=start
function romanToInt(s: string): number {
    let m = new Map()
    m.set("M", 1000)
    m.set("CM", 900)
    m.set("D", 500)
    m.set("CD", 400)
    m.set("C", 100)
    m.set("XC", 90)
    m.set("L", 50)
    m.set("XL", 40)
    m.set("X", 10)
    m.set("IX", 9)
    m.set("V", 5)
    m.set("IV", 4)
    m.set("I", 1)
    if (s.length == 1)
        return m.get(s[0])
    let res = 0
    for (let i = 0; i < s.length - 1; i++) {
        let key = s[i] + s[i + 1]
        if (m.get(key) != null)
            res += m.get(key), i++
        else
            res += m.get(s[i])
    }
    if (m.get(s[s.length - 2] + s[s.length - 1]) == null)
        res += m.get(s[s.length - 1])
    return res

};
// @lc code=end

File: c++/leetcode/13.罗马数字转整数.cpp
/*
 * @lc app=leetcode.cn id=13 lang=cpp
 *
 * [13] 罗马数字转整数
 */
#include <string>
#include <unordered_map>
using namespace std;
// @lc code=start
class Solution {
public:
    unordered_map<char, int> dict = {
           {'I', 1},
           {'V', 5},
           {'X', 10},
           {'L', 50},
           {'C', 100},
           {'D', 500},
           {'M', 1000},
    };
    int romanToInt(string s) {
        int ans = 0;
        for (int i = 0;i < s.size();i++) {
            if (i < s.size() - 1 && dict[s[i]] < dict[s[i + 1]])ans -= dict[s[i]];
            else ans += dict[s[i]];
        }
        return ans;
    }
};
// @lc code=end

